oeps/oep-0007.rst

```diff
@@ -0,0 +1,462 @@
+should initially support versions 3.5 and up.  Libraries that support Python 2
+should only support version 2.7.  Future version deprecations are outside the
+scope of this document.
+
+The end goal is for all of our services to be deployed under Python 3 or
+replaced by newer services before the April 2020 end of support.
+
+
+Migrating code
+++++++++++++++
+
+All new code in Python 2 codebases should be written to be compatible with Python 3.  There
+are two major libraries to help work with compatibility differences.  Both have
+their uses, so rather than mandating one over the other, we offer situations
+where each one is preferable.  The two options are `six`, and `future`.
+
+Six is a simple compatibility library.  If something works differently in
+Python 2 and and Python 3, you import six, and use functions that it provides
+rather than the ambiguous code.  Rather than using Python 3's `str` class,
+which would be interpreted as a bytestring in Python 2, you use
+`six.text_type`.  Instead of calling `d.items()` on a dictionary, you call
+`six.viewitems(d)`.  Stdlib imports that have changed can be imported from
+`six.moves.*`.  The compatibility layer is intrusive throughout the code base,
+but it is explicit, easy to find, and easy to control.
+
+Future, on the other hand, tries to allow developers to write Python 3 native
+code as much as possible, while restricting most compatibility boilerplate to
+the file's imports.  To get an unambiguous text object, you do add `from
+builtins import str`.  Most stdlib imports can just be imported using their
+Python 3 name, and future handles adding compatibility shims in the right place
+for Python 2.  Dictionaries still need to use a `future.utils.viewitems(d)`
+shim, as there's no way to override dict literals.  The goal is to make the
+code look as much as possible like native Python 3 code, and to make the
+compatibility layer easy to remove when dropping Python 2 support.  The
+downside is that it is not always obvious when future is being used, which
+could lead to hard-to-debug issues when maintaining a cross-compatible future
+codebase.
+
+Our general recommendation is to use future if a repository can reasonably be
+converted and drop Python 2 support within a short time frame (roughly two
+months).  If the project cannot be converted quickly, or will need to maintain
+support for Python 2 for a while after conversion, six is a better option, as
+the explictness of the compatibility code will make it easier to find
+compatibility issues, and to avoid introducing new issues.
+
+Third-party dependencies
+------------------------
+
+Before converting a codebase to Python 3, we need to make sure the code we
+depend on will also support Python 3.  We can make a rough check for
+third-party library support using the [Can I Use Python 3 website or
+library](https://caniusepython3.com/).  This will sometimes show inaccurate
```

Comment:
This link and 5 others below use Markdown formatting, but this is a reStructuredText document.  They need to be updated to render correctly in the output.